### PR TITLE
feat(dgw): remux webm files when video recording ends

### PR DESCRIPTION
**Pull request by irvingoujAtDevolution**

### CODE REVIEW

File: devolutions-gateway/Cargo.toml [hunk 1]
```diff
@@ -88,7 +88,7 @@ dlopen = "0.1"
 dlopen_derive = "0.1"
 
 # Video processing for session recording
-cadeau = { version = "0.3", features = ["dlopen"] }
+cadeau = { path = "../../cadeau/rust/cadeau", features = ["dlopen"] }
```
Comment (irvingoujAtDevolution) [inline devolutions-gateway/Cargo.toml:91]: Notice: This will and need to be updated with new release of cadeau

File: devolutions-gateway/src/service.rs [hunk 2]
```diff
@@ -97,7 +97,7 @@ impl GatewayService {
             match result {
                 Ok(_) => info!("XMF native library loaded and installed"),
                 Err(error) => warn!(
-                    %error,
+                    ?error,
```
Comment (CBenoit) [inline devolutions-gateway/src/service.rs:100]: question: Why did you change that? IIRC there is no reason to use the Debug representation at this place.

File: devolutions-gateway/Cargo.toml (hunk 1 above)
Comment (CBenoit) [reply devolutions-gateway/Cargo.toml:91]: I published cadeau v0.4.0

File: devolutions-gateway/src/recording/muxer.rs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+use anyhow::Context;
+
+pub async fn remux(input: impl AsRef<Path>) -> anyhow::Result<()> {
+    let path = Arc::new(input.as_ref().to_owned());
+    // Could potentially be a very slow operation
+    tokio::task::spawn_blocking(|| remux_impl(path)).await??;
+    Ok(())
+}
+
+fn remux_impl(input: Arc<PathBuf>) -> anyhow::Result<()> {
```
Comment (CBenoit) [inline devolutions-gateway/src/recording/muxer.rs:16]: suggestion: You can take a `PathBuf` by move instead of using `Arc`. There is no need for shared ownership.

File: devolutions-gateway/src/recording/muxer.rs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    std::thread::sleep(Duration::from_secs(1));
```
Comment (CBenoit) [inline devolutions-gateway/src/recording/muxer.rs:17]: question: What is the reason for this sleep? If we absolutely need to wait for one second, use an async sleep from tokio outside of the blocking thread. A thread pool is used under the hoods, so we should avoid this kind of pattern when possible.

File: devolutions-gateway/src/recording/muxer.rs
Comment (CBenoit) [inline devolutions-gateway/src/recording/muxer.rs:1]: nitpick: I don’t think it’s worth creating a new module just for the `remux` function. For so few, you can put local `use`s and the helper function directly inside the function if you want some encapsulation.

See 

```rust
pub async fn remux(input_path: Utf8PathBuf) -> anyhow::Result<()> {
    use std::time::Duration;

    // We need to sleep because […]
    tokio::time::sleep(Duration::from_secs(1)).await;

    // CPU-intensive operation potentially lasting much more than 100ms.
    tokio::task::spawn_blocking(move || remux_impl(input_path)).await??;

    return Ok(());

    fn remux_impl(input_path: Utf8PathBuf) -> anyhow::Result<()> {
        /* cut */
    }
}

```

File: devolutions-gateway/src/recording/muxer.rs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            "failed to remux file: {}, output: {}",
+            input_to_remux.display(),
+            output.display(),
+        )
+    })?;
+
+    std::fs::copy(output.as_path(), input_to_remux).with_context(|| "failed to copy remuxed file")?;
+    std::fs::remove_file(output).with_context(|| "failed to remove temporary file")?;
+    debug!(?input_to_remux, "Successfully Remuxed file");
+    Ok(())
```
Comment (CBenoit) [inline devolutions-gateway/src/recording/muxer.rs:42]: suggestion: You can simplify this code a lot by using `Utf8PathBuf`, the `parent` method, `std::fs::rename` a a few intermediate variables. Something like this:

```rust
pub async fn remux(input_path: Utf8PathBuf) -> anyhow::Result<()> {
    use std::time::Duration;

    // We need to sleep because […]
    tokio::time::sleep(Duration::from_secs(1)).await;

    // CPU-intensive operation potentially lasting much more than 100ms.
    tokio::task::spawn_blocking(move || remux_impl(input_path)).await??;

    return Ok(());

    fn remux_impl(input_path: Utf8PathBuf) -> anyhow::Result<()> {
        let input_file_name = input_path
            .file_name()
            .ok_or(anyhow::anyhow!("input file has no file name"))?;

        let remuxed_file_name = format!("remuxed_{input_file_name}");

        let output_path = input_path
            .parent()
            .context("failed to retrieve parent folder")?
            .join(remuxed_file_name);

        cadeau::xmf::muxer::webm_remux(&input_path, &output_path)
            .with_context(|| format!("failed to remux file {input_path} to {output_path}"))?;

        std::fs::rename(&output_path, &input_path).context("failed to override remuxed file")?;

        debug!(%input_path, "Successfully remuxed video recording");

        Ok(())
    }
}
```

File: devolutions-gateway/src/recording/mod.rs [hunk 6]
```diff
@@ -614,6 +631,13 @@ async fn recording_manager_task(
                             id,
                         });
 
+                        if RecordingFileType::WebM == file_type {
+                            tokio::spawn(async move {
+                                if let Err(e) = remux(recording_file).await {
+                                    error!(error = format!("{e:#}"), "remux");
+                                }
+                            });
+                        }
```
Comment (CBenoit) [inline devolutions-gateway/src/recording/mod.rs:640]: suggestion: Move that into `handle_disconnect`, and use the `current_file.file_name` instead.

```rust
            let recording_file_path = ongoing
                .manifest_path
                .parent()
                .expect("a parent")
                .join(&current_file.file_name);

            debug!(path = %ongoing.manifest_path, "Write updated manifest to disk");

            ongoing
                .manifest
                .save_to_file(&ongoing.manifest_path)
                .with_context(|| format!("write manifest at {}", ongoing.manifest_path))?;

            if recording_file_path.extension() == Some(RecordingFileType::WebM.extension()) {
                tokio::spawn(async move {
                    if let Err(e) = remux(recording_file_path).await {
                        error!(error = format!("{e:#}"), "Remux operation failed");
                    }
                });
            }

```

File: devolutions-gateway/src/recording/mod.rs [hunk 7]
```diff
@@ -176,6 +182,8 @@ enum RecordingManagerMessage {
     },
     Disconnect {
         id: Uuid,
+        file_type: RecordingFileType,
+        recording_file: Utf8PathBuf,
```
Comment (CBenoit) [inline devolutions-gateway/src/recording/mod.rs:186]: suggestion: By applying the other suggestions, this becomes unused.
```suggestion
```

File: devolutions-gateway/src/recording/muxer.rs (hunk 4 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/recording/muxer.rs:17]: Sorry, this was for debugging!

File: devolutions-gateway/src/recording/muxer.rs
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/recording/muxer.rs:1]: updated

File: devolutions-gateway/src/recording/muxer.rs (hunk 5 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/recording/muxer.rs]: updated

File: devolutions-gateway/src/recording.rs [hunk 8]
```diff
@@ -676,3 +690,32 @@ async fn recording_manager_task(
 
     Ok(())
 }
+
+pub async fn remux(input_path: Utf8PathBuf) -> anyhow::Result<()> {
+    // CPU-intensive operation potentially lasting much more than 100ms.
+    tokio::task::spawn_blocking(move || remux_impl(input_path)).await??;
```
Comment (CBenoit) [inline devolutions-gateway/src/recording.rs:696]: issue: Check that Cadeau was loaded before using it.
```suggestion
    if cadeau::xmf::is_init() {
        // CPU-intensive operation potentially lasting much more than 100ms.
        tokio::task::spawn_blocking(move || remux_impl(input_path)).await??;
    }
```

File: devolutions-gateway/src/recording.rs [hunk 9]
```diff
@@ -5,7 +5,7 @@ use std::path::Path;
 use std::pin::pin;
 use std::sync::Arc;
 
-use anyhow::Context as _;
+use anyhow::Context;
```
Comment (CBenoit) [inline devolutions-gateway/src/recording.rs:8]:
```suggestion
use anyhow::Context as _;
```

**Review by CBenoit — APPROVED** [review]
LGTM!
